### PR TITLE
fix: we now only add `-g` flag to Java source types

### LINKED ISSUE #1656: Running groovy scripts on windows fails on compiling (-g The system cannot find the file specified)
**Describe the bug**
Trying to run the hello.groovy script on windows (both git bash & windows cmd)

C:\work\misc\groovy-scratch\src>jbang --verbose test.groovy
[jbang] [3:013] jbang version 0.109.0
[jbang] [3:278] Resolving resource ref: test.groovy
[jbang] [3:328] Resolved resource ref as: test.groovy (cached as: C:\work\misc\groovy-scratch\src\test.groovy)
[jbang] [3:559] Build required as C:\Users\SamuliSaarinen\.jbang\cache\jars\test.groovy.2a85157bad3bb835195b0c4a19ed409de05d6d16560310c04b96e43ebeaf8b34\test.jar not readable or not found.
[jbang] [3:579] Resolving artifact(s): org.codehaus.groovy:groovy:3.0.9
[jbang] [3:596] Repositories: mavencentral=https://repo1.maven.org/maven2/
[jbang] [3:833] Resolved artifact(s) from cache: org.codehaus.groovy:groovy:jar:3.0.9=C:\Users\SamuliSaarinen\.m2\repository\org\codehaus\groovy\groovy\3.0.9\groovy-3.0.9.jar
[jbang] [4:129] Building jar for test.groovy...
[jbang] [4:129] Compile: C:\Users\SamuliSaarinen\.jbang\cache\groovycs\3.0.9\groovy-3.0.9\bin\groovyc.bat -g -classpath C:\Users\SamuliSaarinen\.m2\repository\org\codehaus\groovy\groovy\3.0.9\groovy-3.0.9.jar -d C:\Users\SamuliSaarinen\.jbang\cache\jars\test.groovy.2a85157bad3bb835195b0c4a19ed409de05d6d16560310c04b96e43ebeaf8b34\classes C:\work\misc\groovy-scratch\src\test.groovy
[jbang] [4:151] Looking for JDK: 0
[jbang] [4:228] Using JDK: 17 (17.0.8+7, current, C:\Users\SamuliSaarinen\.jbang\cache\jdks\17)
org.codehaus.groovy.control.MultipleCompilationErrorsException: startup failed:
-g: -g (The system cannot find the file specified)

1 error

[jbang] [14:421] Deleting folder C:\Users\SamuliSaarinen\.jbang\cache\jars\test.groovy.2a85157bad3bb835195b0c4a19ed409de05d6d16560310c04b96e43ebeaf8b34\classes
[jbang] [14:437] Deleting folder C:\Users\SamuliSaarinen\.jbang\cache\jars\test.groovy.2a85157bad3bb835195b0c4a19ed409de05d6d16560310c04b96e43ebeaf8b34\generated
[jbang] [14:437] [ERROR] Error during compile
dev.jbang.cli.ExitException: Error during compile
        at dev.jbang.source.buildsteps.CompileBuildStep.runCompiler(CompileBuildStep.java:145)
        at dev.jbang.source.sources.GroovySource$GroovyAppBuilder$GroovyCompileBuildStep.runCompiler(GroovySource.java:97)
        at dev.jbang.source.buildsteps.CompileBuildStep.runCompiler(CompileBuildStep.java:133)
        at dev.jbang.source.buildsteps.CompileBuildStep.compile(CompileBuildStep.java:117)
        at dev.jbang.source.buildsteps.CompileBuildStep.build(CompileBuildStep.java:56)
        at dev.jbang.source.buildsteps.CompileBuildStep.build(CompileBuildStep.java:41)
        at dev.jbang.source.AppBuilder.build(AppBuilder.java:97)
        at dev.jbang.source.AppBuilder.build(AppBuilder.java:22)
        at dev.jbang.cli.Run.doCall(Run.java:89)
        at dev.jbang.cli.BaseCommand.call(BaseCommand.java:145)
        at dev.jbang.cli.BaseCommand.call(BaseCommand.java:21)
        at picocli.CommandLine.executeUserObject(CommandLine.java:1953)
        at picocli.CommandLine.access$1300(CommandLine.java:145)
        at picocli.CommandLine$RunLast.executeUserObjectOfLastSubcommandWithSameParent(CommandLine.java:2358)
        at picocli.CommandLine$RunLast.handle(CommandLine.java:2352)
        at dev.jbang.cli.JBang$3.handle(JBang.java:148)
        at dev.jbang.cli.JBang$3.handle(JBang.java:143)
        at picocli.CommandLine$AbstractParseResultHandler.execute(CommandLine.java:2179)
        at picocli.CommandLine$RunLast.execute(CommandLine.java:2316)
        at picocli.CommandLine.execute(CommandLine.java:2078)
        at dev.jbang.Main.main(Main.java:14)

**Additional context**
Seems like there is a `-g` command line flag added to groovyc but I can't find it in the groovyc documentation and it seems to cause the error. Running the groovyc command without -g runs fine

e.g.
```
C:\Users\SamuliSaarinen\.jbang\cache\groovycs\3.0.9\groovy-3.0.9\bin\groovyc.bat -classpath C:\Users\SamuliSaarinen\.m2\repository\org\codehaus\groovy\groovy\3.0.9\groovy-3.0.9.jar -d C:\Users\SamuliSaarinen\.jbang\cache\jars\test.groovy.2a85157bad3bb835195b0c4a19ed409de05d6d16560310c04b96e43ebeaf8b34\classes C:\work\misc\groovy-scratch\src\test.groovy
```

### PR DESCRIPTION
**Pull request by quintesse**
Before it would get added to Groovy and Kotlin as well.

Fixes #1656

### CODE REVIEW

File: src/main/java/dev/jbang/source/sources/JavaSource.java [hunk 1]
```diff
@@ -25,7 +27,9 @@ protected JavaSource(ResourceRef ref, String script, Function<String, String> re
 
 	@Override
 	protected List<String> getCompileOptions() {
-		return tagReader.collectOptions("JAVAC_OPTIONS", "COMPILE_OPTIONS");
+		List<String> gopts = Collections.singletonList("-g");
```
Comment (maxandersen) [inline src/main/java/dev/jbang/source/sources/JavaSource.java:30]: not really a fan of adding this here - won't this mean that if you have 10 source files you get 10 x "-g" ?

Comment (quintesse) [reply src/main/java/dev/jbang/source/sources/JavaSource.java:30]: Possibly, yes. For now I just don't know where else to put it :-/
For some time now I've thought that there should probably be a class for each kind of project, eg JavaProject, KotlinProject, etc. Then we'd have a proper place to put things like that.

Comment (quintesse) [reply src/main/java/dev/jbang/source/sources/JavaSource.java:30]: We can of course leave it where it is and surround it with an instanceof test to see if the main source file is JavaSource or not.

Comment (maxandersen) [reply src/main/java/dev/jbang/source/sources/JavaSource.java:30]: maybe have a `LinkedSet contributeGlobalCompileOptions();` which allows a Source to add to a set that gets added first?

**Comment by quintesse** [comment]
I tried coming up with a nice solution, by introducing different kinds of Projects but it got a little out of hand. So I need to rethink that at some point. So for now I'm using an ugly `instanceof`, but at least it works.

**Review by maxandersen — APPROVED** [review]
not great but better than annoying fail.

**Pull request merged**